### PR TITLE
fix: check all locations if in is not specified.

### PR DESCRIPTION
**Pull request by nikli2009**
## Description

The `default` option doesn't work as expected when the field is in `req.query` and `in` is not specified.

## To-do list

- [ ] I have added tests for what I changed.

- [ ] This pull request is ready to merge.

### CODE REVIEW

File: src/context.ts [hunk 1]
```diff
@@ -43,12 +43,12 @@ export class Context {
...
 
         // #331 - When multiple locations are involved, all of them must pass the validation.
-        // If none of the locations contain the field, we at least include one for error reporting.
+        // If none of the locations contain the field, we should check all locations since we are not sure which location the consumer uses.
         // #458, #531 - Wildcards are an exception though: they may yield 0..* instances with different
         // paths, so we may want to skip this filtering.
         if (instances.length > 1 && locations.length > 1 && !group.includes('*')) {
           const withValue = instances.filter(instance => instance.value !== undefined);
-          return withValue.length ? withValue : [instances[0]];
+          return withValue.length ? withValue : instances;
```
Comment (fedeci) [inline src/context.ts:51]: This imply that even locations where `!Object.prototype.hasOwnProperty(location, param)` get the default value added🤔

Comment (nikli2009) [reply src/context.ts:51]: Hi buddy, thanks for the review. :D 
Just for clarification, the line I changed (Line 51) used to make the `default` only works for `req.body` (which is the first item of `instances`), and I thought maybe it would probably makes more sense to make it works for all `location`s when consumer hasn't set any values for `in`.

So in the commit, I changed to `instances` and this makes the `default` feature in `checkSchema` to be more consistent with the other scenario that this lib currently supports already. [here](https://express-validator.github.io/docs/sanitization-chain-api.html#defaultdefault_value)

If we set up like 
```
app.use('/user', query('greeting').default('hiya'), userController);
```
and send request `GET /user`, the `req.query` will be transformed into `{ greeting: 'hiya' }`

So in short, like you commented, `!Object.prototype.hasOwnProperty(location, param)` does get the default value added... and this behavior is consistent with the other API [here](https://express-validator.github.io/docs/sanitization-chain-api.html#defaultdefault_value)

Comment (fedeci) [reply src/context.ts:51]: Well #1070 is actually not related to this PR as written in the issue.
However the first problem that comes to my mind is that line 51 is not only used by `.default` and is shared across all the other validators and sanitizers. The fix may be quite though sadly :(

Comment (nikli2009) [reply src/context.ts:51]: Got you, thanks for maintain the lib ! Glad to add more tests if can help save some verification effort (maybe another PR).

File: src/middlewares/schema.spec.ts [hunk 2]
```diff
@@ -183,12 +183,29 @@ describe('on each field', () => {
...
+    expect(context.getData().every(instance => instance.value === 10)).toBe(true);
+    expect(req.query.undef).toBe(10);
+  });
 });
 
 describe('on schema that contains fields with bail methods', () => {
   it('stops validation chain with only one error', async () => {
     const schema = checkSchema({
       foo: {
+        in: ['params'],
```
[the remark this answers is not in the document]
Comment (nikli2009) [reply src/middlewares/schema.spec.ts:224]: Just update the tests and one thing I noticed is: 
When `in` is not specified then check all locations by default, user will be getting 5 errors from 5 default locations. However this means if use `exists` + `bail` method and the field is not in any location of the req object,  user will still get 5 errors.

[the remark this answers is not in the document]
Comment (fedeci) [reply src/middlewares/schema.spec.ts:224]: Was it like that before?

**Comment by fedeci** [comment]
Hey @nikli2009 can you add a few more tests to check if #1010 is fixed by this?

**Comment by nikli2009** [comment]
> Hey @nikli2009 can you add a few more tests to check if #1010 is fixed by this?

hey man, gladly ! checking today

File: src/middlewares/schema.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikli2009) [reply src/middlewares/schema.spec.ts:224]: Just updated the tests along with some explanation comments, it used to only generate `1` error after checking `req.body`, however as location is not specified, we can't actually know which location user intend to check.

**Comment by nikli2009** [comment]
> Hey @nikli2009 can you add a few more tests to check if #1010 is fixed by this?

Test added, could you help review and let me know if need to add more

File: src/context.ts [hunk 3]
```diff
@@ -43,12 +43,12 @@ export class Context {
         const locations = _.uniqBy(instances, 'location');
 
         // #331 - When multiple locations are involved, all of them must pass the validation.
-        // If none of the locations contain the field, we at least include one for error reporting.
+        // If `in` option is not specified, try to check all locations by default.
```
Comment (fedeci) [inline src/context.ts:46]:
```suggestion
        // checkSchema: If `in` option is not specified, try to check all locations by default.
```

File: src/middlewares/schema.spec.ts [hunk 4]
```diff
@@ -200,7 +233,57 @@ describe('on schema that contains fields with bail methods', () => {
...
+          },
+        },
+      },
+    });
+
+    const { context } = await schema[0].run({ query: {} });
+    // when location is not specified, checking each location one by one,
+    // if the expected field does not show up at all in any location, then there will be 5 errors.
+    // finally bails.
+    expect(context.errors).toHaveLength(allLocations);
```
Comment (fedeci) [inline src/middlewares/schema.spec.ts:260]: nit: we can hardcode `allLocations` in the tests since locations will unlikely change in length.

File: src/middlewares/schema.spec.ts [hunk 5]
```diff
@@ -183,12 +183,45 @@ describe('on each field', () => {
...
+          options: 10,
+        },
+      },
+    })[0];
+
+    const req = { query: { undef: undefined } };
+    const { context } = await chain.run(req);
+
+    expect(context.getData().every(instance => instance.value === 10)).toBe(true);
+    expect(req.query.undef).toBe(10);
```
Comment (fedeci) [inline src/middlewares/schema.spec.ts:200]: Can we check that also `req.body.undef`, ... have value of `10`. We will fix this testing problem with fixtures soon :)

**Review by fedeci — APPROVED** [review]
Some nitty comments but look fine. Thanks!
/cc @gustavohenke patch or minor? It actually corrects a bug(?) but may break something🤔

File: src/middlewares/validation-chain-builders.spec.ts [hunk 6]
```diff
@@ -48,6 +48,32 @@ describe('buildCheckFunction()', () => {
...
   });
+
+  describe('when at least 1 locations specified', () => {
+    // for #1010
+    it('should apply default value to each specified location if default option is not empty', async () => {
+      const custom = buildCheckFunction(['cookies', 'headers']);
+      const chain = custom('foo').default(10);
+
+      // reset req
+      req = {
```
[the remark this answers is not in the document]
Comment (nikli2009) [reply src/middlewares/validation-chain-builders.spec.ts:59]: good one! will update

File: src/middlewares/schema.spec.ts (hunk 4 above)
Comment (nikli2009) [reply src/middlewares/schema.spec.ts:260]: sure 👌

**Comment by nikli2009** [comment]
Tests updated + comment squashed 🥪🥪 .

**Review by gustavohenke — CHANGES_REQUESTED** [review]
This is a breaking change.
For everybody using `check()` instead of the specialised functions, this can be a bit wild...

Test code:
```js
const req = {};
check('foo').isNumeric().run(req).then(() => {
  console.log(validationResult(req).array());
});
```

Output:

```js
[
  {
    value: undefined,
    msg: 'Invalid value',
    param: 'foo',
    location: 'body'
  },
  {
    value: undefined,
    msg: 'Invalid value',
    param: 'foo',
    location: 'cookies'
  },
  {
    value: undefined,
    msg: 'Invalid value',
    param: 'foo',
    location: 'headers'
  },
  {
    value: undefined,
    msg: 'Invalid value',
    param: 'foo',
    location: 'params'
  },
  {
    value: undefined,
    msg: 'Invalid value',
    param: 'foo',
    location: 'query'
  }
]
```

---

Looking at #1071, it's conceptually wrong that you don't pass `in` to the schema, and expect a certain request location to have the value you wanted.
You should probably use `matchedData()` instead.

But then, a few things I'm wondering:
- would it makes sense to differentiate between "all/multiple locations" vs "any location", so to appease to those that want the convenience of `check`?
- should we verify if the user actually tried to pass the validated field somewhere -- e.g. `/some?query` vs `/some` -- and prefer handling `req.query` instead of `req.body`?
